EA-207: Implement an API for Fetching Visits with Notes and Diagnosis

jayasanka-sack opened the pull request:
Ticket: https://openmrs.atlassian.net/browse/EA-207

We needed to create a new API that retrieves visits along with their associated notes and diagnosis to optimize the current approach. The existing API call we are using to fetch past visits is expensive and returns more data than necessary for the initial UI rendering. The goal is to retrieve only the relevant information (visits, notes, diagnosis) to improve performance and load times.

mogoodrich commented:
@jayasanka-sack  can you write up an EMR-API ticket in Jira describing the use cases on what this PR is intended to accomplish?  thanks, fyi @mseaton

jayasanka-sack commented:
Hey @mogoodrich The ticket is: https://openmrs.atlassian.net/browse/EA-207

But this PR is still under WIP. I'll let you know when it is ready for review.

mogoodrich reviewed (changes requested):
Thanks @jayasanka-sack !  Good to see we are adding methods to the EMR API to support performant queries in O3.

See my comments and let me know if you have questions... also, can you add a link to this PR to the ticket?  

Also, it looks like this is really just returning "Visits With Diagnoses" (not "Visits with Notes and Diagnsoes")... which is fine, just wanted to confirm it is what you were looking for.

also fyi @mseaton

mogoodrich commented on api/src/main/java/org/openmrs/module/emrapi/db/VisitDAO.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.util.List;
+
+public interface VisitDAO {
```
I think we may want to call with EmrApiVisitDAO to distinguish it from the Core one?  Thoughts @mseaton  @ibacher ?

ibacher replied:
Agreed.

mseaton replied:
Do we really need an entirely new DAO class we have to wire up and maintain?  Can't we just move to having one DAO for the entire module?

mogoodrich commented on api/src/main/java/org/openmrs/module/emrapi/visit/VisitWithDiagnosesService.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.util.List;
+
+public interface VisitWithDiagnosesService {
```
Maybe call this EmrApiVisitService (and we can use it for other Visit-related services we add to EMR API)?  Thoughts @mseaton  @ibacher ?

ibacher replied:
Yeah, that's a very good idea!

mseaton replied:
Yes!  Let's move towards having fewer distinct DAOs and Services in this module

mseaton commented on api/src/main/java/org/openmrs/module/emrapi/db/HibernateVisitDAO.java:
```diff
@@ -0,0 +1,74 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Repository
```
I have never seen this annotation actually used in this module before.  Are we sure this works?  I'd prefer we stick with existing conventions within the module for wiring things up, and then do a change to everything if we want to update to a new mechanism.

mseaton commented on api/src/main/java/org/openmrs/module/emrapi/db/HibernateVisitDAO.java:
```diff
@@ -0,0 +1,74 @@
+@Repository
+public class HibernateVisitDAO implements VisitDAO {
+
+    @Autowired
```
Usually we do not autowire into a DAO AFAIK, but we use xml.  There are definitely performance issues to using Autowired in services, and I'm not sure if those carry over to DAOs.

mseaton commented on omod/pom.xml:
```diff
@@ -149,6 +149,12 @@
+            <artifactId>junit-jupiter</artifactId>
+            <version>RELEASE</version>
+            <scope>test</scope>
+        </dependency>
```
We either should choose to use Junit 4 or Junit 5 for the entire module.  I don't want us to have to try to support both in one module, there are too many annoying gotchas.

mogoodrich replied:
Yes, if it's problematic to support both JUnit 4 and JUnit 5 in a single module, let's just go with JUnit 4 until we feel motivated to update all the tests.

mseaton commented on omod/src/main/java/org/openmrs/module/emrapi/web/controller/VisitController.java:
```diff
@@ -0,0 +1,57 @@
+    VisitWithDiagnosesService visitWithDiagnosesService;
+    
+    @RequestMapping(method = RequestMethod.GET, value = "/rest/**/emrapi/patient/{patientUuid}/visit")
+    public ResponseEntity<?> getVisitsByPatientId(
```
And the endpoint should not be `/.../visit`, but should reflect that it is a `/.../visitWithDiagnoses` or something

jayasanka-sack commented:
Appreciate the feedback @mogoodrich and @mseaton! I’ll take a look and follow up with you.

jayasanka-sack commented on omod/src/main/java/org/openmrs/module/emrapi/web/controller/VisitController.java:
```diff
@@ -0,0 +1,57 @@
+    public ResponseEntity<?> getVisitsByPatientId(
+            HttpServletRequest request,
+            HttpServletResponse response,
+            @PathVariable String patientUuid) {
```
I’m guessing it needs to be called with ?patient={uuid} but it’s returning a 500 with no body, and I’m having a hard time debugging it. Can you point me to where the conversion happens? I noticed a couple of other endpoints which use the same, but I couldn’t find any tests for them or any part of the frontend that calls them.

ibacher replied:
Unfortunately, it's part of the [UI framework](https://github.com/openmrs/openmrs-module-uiframework/blob/1f88e1892b9becddf44c9f6666e73947bcb0aad9/api/src/main/java/org/openmrs/ui/framework/converter/StringToPatientConverter.java#L8), which is almost certainly not the right place for it. I think you could also do `@PathVariable("patientUuid") Patient patient` with the right converter.

ibacher replied:
Did you try adding the UIFramework module?

ibacher replied:
@mogoodrich @mseaton Do you think there's any objection to migrating the various converters out of the UI framework and into somewhere else? (My two initial thoughts are either: add them to EMR API or add them to the REST module; they don't _quite_ fit in the REST module, but it's actually a helpful thing to have).

mseaton replied:
No real objection, but also no need to get bogged down on this.  Just accept a String parameter (which could be a uuid or primary key id), and convert it to a patient in the controller...

mogoodrich replied:
Huh, so things like the ActiveVisitController have been dependent on UI Framework module all along, we just never realized it because we've generally always been running EMR API and UI Framework together?

No strong objection to move them... I think I'd have slight preference for moving them to EMR API instead of REST Web Services, but haven't thought it all through.  We also could potentially add them to Core? (Though of course the adoption time would be slower that way).

mogoodrich replied:
Also, it looks like REST WS may have it's own converters, fwiw...

ibacher replied:
REST has [it's own Converter concept](https://github.com/openmrs/openmrs-module-webservices.rest/blob/f3d8915f3990c9e65e49e0f33495b497b6d2ac81/omod-common/src/main/java/org/openmrs/module/webservices/rest/web/resource/api/Converter.java#L20), but it's actually distinct from the Spring converters, I think.

ibacher replied:
> We also could potentially add them to Core? (Though of course the adoption time would be slower that way)

Yeah, I'd like to migrate them to core eventually, but we just released 2.7.0, so it's not the _best_ time to do that. Easy enough to add to a module and then add an `@OpenmrsProfile` to exclude them once they're available in core.

jayasanka-sack commented on api/src/main/java/org/openmrs/module/emrapi/db/HibernateVisitDAO.java:
```diff
@@ -0,0 +1,74 @@
+
+    public List<VisitWithDiagnoses> getVisitsByPatientId(Patient patient, int startIndex, int limit) {
+
+        String visitNoteEncounterTypeUuid = "d7151f82-c1f3-4152-a605-2f9ea7414a79";
```
We’re including this to get only note encounters with the visit, but we also need visits that don’t have encounters. I’ve updated the query with a left join and modified the test accordingly.

jayasanka-sack commented:
I still have to address some comments. I'll get them fixed and get back to you.

mogoodrich reviewed (commented):
Sorry, juts realized I had a couple "pending" comments that I never submitted.

jayasanka-sack commented on api/src/test/resources/metadata/pastVisitSetup.xml:
```diff
@@ -0,0 +1,35 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Both files are used in tests in both modules.

mogoodrich replied:
What is the distinction between them?

ibacher replied:
Yeah the organization is a little weird. It looks like there may have been some convention for all the metadata to go into a file like this and the data into a file like the other one, but both of them mix data and metadata, so it's confusing why we need the two.

jayasanka-sack replied:
My bad! This is an extra file that wasn’t needed. By the way, about the XML file that was used in both modules’ tests, I ended up reusing the same one using the test jar.

mogoodrich reviewed (commented):
Thanks @jayasanka-sack , getting close!  See my updated thoughts/questions.

jayasanka-sack commented on api/src/main/java/org/openmrs/module/emrapi/db/EmrApiDAOImpl.java:
```diff
@@ -55,4 +67,54 @@ public <T> List<T> executeHqlFromResource(String resource, Map<String, Object> p
+      String visitNoteEncounterTypeUuid = emrApiProperties.getVisitNoteEncounterType().getUuid();
+      
+      String hqlVisit="SELECT DISTINCT v FROM Visit v " +
+              "LEFT JOIN FETCH v.encounters enc " +
```
Yes, you’re right! 

I used the FETCH syntax to load encounters eagerly instead of lazily.

mogoodrich replied:
fyi @mseaton  @ibacher

ibacher reviewed (commented):
Thanks @jayasanka-sack! A few comments and suggestions.

mogoodrich commented:
Sorry @jayasanka-sack  I haven't been following this... is this ready for re-review?

ibacher commented:
@mogoodrich Yes, it is.

mseaton commented on omod/src/main/resources/moduleApplicationContext.xml:
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This is unnecessary.  Just use Autowired on the emrApiVisitService property on the controller.  It is already an annotated controller, so does not need to be wired in like this, does it?

mogoodrich commented on api/src/main/java/org/openmrs/module/emrapi/db/EmrApiDAOImpl.java:
```diff
@@ -55,4 +67,54 @@ public <T> List<T> executeHqlFromResource(String resource, Map<String, Object> p
       return executeHql(hql, parameters, clazz);
    }
+   
+   public List<VisitWithDiagnoses> getVisitsWithNotesAndDiagnosesByPatient(Patient patient, int startIndex, int limit) {
```
@ibacher  are you referring to the "VisitWithDiagnoses" type (and the fact that "Notes" is missing?)  This does seem kind of weird, though, the object is literally just a "VisitWithDiagnoses" object, but (if I understand this correctly) when we populate this Visit in the DAO we *only* populate it with the Visit Note encounters associated with it, not all the encounters associated with it.

mogoodrich commented on api/src/test/java/org/openmrs/module/emrapi/db/EmrApiDaoTest.java:
```diff
@@ -0,0 +1,90 @@
+        assert firstVisitDiagnoses.size() == 3;
+
+        for (Encounter encounter : firstVisitEncounters) {
+            assert encounter.getEncounterType().getUuid().equals(visitNoteEncounterTypeUuid);
```
Following up on my comment from above, I'm a little confused by this, as it looks like in the test data below this visit has two encounters but one is *not* a visit note?

jayasanka-sack replied:
Yeah, the test data includes other encounters too, but this check ensures that only visit encounters are returned. That said, it's slightly better now with the new property for visits.

mogoodrich reviewed (changes requested):
I have some comments, but I have a bigger question about the design, apologies for not raising it earlier.

If I'm understanding, the REST endpoint returns something in the format:

```
[
  {
      "visit": ...
      "diagnoses": ...
  },
  {
      "visit": ...
      "diagnoses": ...
  }
]
```

... but via the magic of Hibernate someone the only encounters on those visits objects are ones that are visit notes?  Am I understanding that right and are we sure that works?  Beyond that, I think it's confusing for the consumer because the assumption is that the visit contains *all* the encounters for patient.  I'd prefer to return something in this format:

```
[
  {
      "visit": ...
      "diagnoses": ...
      "visitNotes": ....
  },
  {
      "visit": ...
      "diagnoses": ...
      "visitNotes": ....
  }
]
```

Am I understanding this correctly? Apologies for not flagging this sooner.  Also fyi @mseaton  @ibacher

ibacher commented:
Oh! I see I missed that we're fetching the encounters with the visit note type.!

jayasanka-sack commented:
Thanks a bunch for your feedback, @ibacher, @mogoodrich, @mseaton! I’ll take a look at these comments and get back to you soon.

jayasanka-sack commented on api/src/main/java/org/openmrs/module/emrapi/db/EmrApiDAOImpl.java:
```diff
@@ -55,4 +67,54 @@ public <T> List<T> executeHqlFromResource(String resource, Map<String, Object> p
+              "LEFT JOIN enc.encounterType et " +
+              "WHERE v.patient.id = :patientId " +
+              "AND (et.uuid = :encounterTypeUuid OR enc IS NULL) " +
+              "ORDER BY v.startDatetime DESC";
```
Yeah, I removed the sorting from the stream as database level sorting is generally faster.

jayasanka-sack replied:
Fixed with https://github.com/openmrs/openmrs-module-emrapi/pull/250/commits/3643d16203cc6ccd115a6e335e9968e89c545dc3

jayasanka-sack commented on api/src/main/java/org/openmrs/module/emrapi/visit/EmrApiVisitServiceImpl.java:
```diff
@@ -0,0 +1,31 @@
+    @Override
+    public List<VisitWithDiagnoses> getVisitsWithNotesAndDiagnosesByPatient(String patientUuid, int startIndex, int limit) {
+
+        Patient patient = patientService.getPatientByUuid(patientUuid);
```
Patient UUID is a path parameter, which cannot be empty. If it is empty, Spring will treat `//` as a different path and will not match it to the controller, resulting in a 404 Not Found error

jayasanka-sack replied:
fixed with https://github.com/openmrs/openmrs-module-emrapi/pull/250/commits/3643d16203cc6ccd115a6e335e9968e89c545dc3

jayasanka-sack commented:
@mogoodrich Your suggestion makes more sense since it prevents confusion where developers might think it includes all encounters. So I added the visitNote property to the `VisitWithDiagnosesAndNotes` class and removed encounters (which are basically visit notes) from visit objects by setting them to an empty hashmap. Is there a better way to completely remove the property from the payload?

jayasanka-sack commented on api/src/main/java/org/openmrs/module/emrapi/visit/EmrApiVisitServiceImpl.java:
```diff
@@ -0,0 +1,31 @@
+        Patient patient = patientService.getPatientByUuid(patientUuid);
+
+        if (patient == null) {
+            throw new ObjectNotFoundException("No patient found with uuid " + patientUuid, Patient.class.getName());
```
Done, and updated the controller to catch it and send a 400 code.

jayasanka-sack commented on api/src/test/java/org/openmrs/module/emrapi/db/EmrApiDaoTest.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    @Test
+    public void shouldFetchVisitsWithNotesAndDiagnosesByPatientId() {
```
According to the my previous implementation, it won't delete other encounters as I didn't explicitly setEncounters later, so that it won't marked as a dirty object. So the encounters won't have an impact event if it is persisted.

However, after the change, now that I set encounters to an empty set later, hibernate will delete the encounters if the visits is re-saved. How about marking the transaction as a read only one? I think this depends on the answer to my question here: https://github.com/openmrs/openmrs-module-emrapi/pull/250#issuecomment-2637004502

mogoodrich replied:
I still think it's better to return visitNotes as a separate field on the result rather than relying on this Hibernate magic to return a visit with only a subset of encounters. @mseaton  @ibacher @jayasanka-sack ?

mogoodrich replied:
Oh sorry, just saw your other comment @jayasanka-sack let me comment on that...

mogoodrich commented:
> @mogoodrich Your suggestion makes more sense since it prevents confusion where developers might think it includes all encounters. So I added the visitNote property to the `VisitWithDiagnosesAndNotes` class and removed encounters (which are basically visit notes) from visit objects by setting them to an empty set. Is there a better way to completely remove the property from the payload?

At the end of the day, this new service method is for use for a new endpoint, correct? If you change that endpoint to only include a REF reference of the encounters, does that fix the performance problems without having to specifically *remove* the encounters property?

jayasanka-sack commented:
@mogoodrich yeah that would also work.  A solution would be:

1. Get visits without filtering encounters
2. Filter note encounters of each visits and include into the property visitNotes

Since visited are converted in default representation (unless specified), encounters will have in REF representation. And notes are converted in default representation too, since they'll converted manually. Using `FETCH` for loading encounters is still a good idea to avoid N+1 query problem.

However, my concern is that if we 100% do not need encounters for this use case, is it worth the extra work? It increases both time complexity (`O(k * n)`) and memory usage where:

- `k` = number of visits per request (constant due to pagination),
- `n` = number of encounters per visit (mostly ≤ 20 but still adds up over many requests).

This might be not noticeable, but when this is on a server the inefficiency scales with increasing traffic.

Perhaps this might not a big deal, but jus wanted to bring it up since the goal is to improve performance of visit fetching, let me know what you think. 

cc: @ibacher @mseaton

jayasanka-sack commented:
btw, it's worth noting that serialization process already has an O(k * n) complexity when converting notes. So it won’t reduce the total time complexity by more than half. But will reduces unnecessary data fetching and memory usage slightly.

mogoodrich commented:
Thanks @jayasanka-sack !  It feels like historically the it's been Hibernate and hydrating Hibernate-managed objects that cause the slowdown, so I was hoping that just having the REF representation of encounters would solve this.  But I can see if you have to fetch all the encounters and then filter them by encounter type that would likely hurt performance.

I'm okay with just removing the encounters from the visit property, but I would think then the visit should not have an "encounter" property at all (as opposed to an empty set, which would suggest that the visit has no encounters) but this might be tricky to do).  Thoughts @ibacher  @mseaton ?

ibacher commented on api/src/main/java/org/openmrs/module/emrapi/visit/VisitWithDiagnosesAndNotes.java:
```diff
@@ -0,0 +1,24 @@
+    
+    private Visit visit;
+    private Set<Diagnosis> diagnoses;
+    private Set<Encounter> visitNotes;
```
For visit notes, we don't even necessarily just need the _encounter_; it's just the individual notes obs withing them that really matter.

jayasanka-sack replied:
Fixed it. Also, used a custom representation to filter out unnecessary data (plus to prevent additional queries)

jayasanka-sack commented:
Update: @ibacher and I had a discussion, and the outcome was to use a custom representation, so that we can eliminate encounters and other irrelevant properties, which also helps us to get rid of additional db queries.

Other changes:
- Fetch only Observation of each note encounters
- Updated queries to filter out voided records
- Update test cases to test voided data

jayasanka-sack commented on api/src/main/java/org/openmrs/module/emrapi/db/EmrApiDAOImpl.java:
```diff
@@ -55,4 +67,58 @@ public <T> List<T> executeHqlFromResource(String resource, Map<String, Object> p
+      
+      String hqlDiagnosis = "SELECT DISTINCT diag FROM Diagnosis diag " +
+              "JOIN diag.encounter e " +
+              "WHERE e.visit.id IN :visitIds";
```
Fixed with https://github.com/openmrs/openmrs-module-emrapi/pull/250/commits/3643d16203cc6ccd115a6e335e9968e89c545dc3

mogoodrich commented:
> Update: @ibacher and I had a discussion, and the outcome was to use a custom representation, so that we can eliminate encounters and other irrelevant properties, which also helps us to get rid of additional db queries during the conversion.
> 
> Other changes:
> 
> * Fetch only Observation of each note encounters
> * Updated queries to filter out voided records
> * Update test cases to test voided data

Sounds good to me, thanks @jayasanka-sack !

ibacher commented on api/src/main/java/org/openmrs/module/emrapi/db/EmrApiDAOImpl.java:
```diff
@@ -55,4 +64,86 @@ public <T> List<T> executeHqlFromResource(String resource, Map<String, Object> p
+      // create VisitWithDiagnosesAndNotes objects
+      List<VisitWithDiagnosesAndNotes> visitWithDiagnosisAndNotes = visits.stream()
+              .map(visit -> {
+                 ArrayList<Obs> notes = visitIdToNotesMap.getOrDefault(visit.getId(), new ArrayList<>());
```
```suggestion
                 ArrayList<Obs> notes = visitIdToNotesMap.getOrDefault(visit.getId(), ArrayList::new);
```

ibacher commented on omod/src/main/java/org/openmrs/module/emrapi/web/controller/VisitController.java:
```diff
@@ -0,0 +1,77 @@
+        List<VisitWithDiagnosesAndNotes> visitsEntries;
+        try {
+            visitsEntries = emrApiVisitService.getVisitsWithNotesAndDiagnosesByPatient(patientUuid, context.getStartIndex(), context.getLimit());
+        }catch (APIException e) {
```
```suggestion
        } catch (APIException e) {
```

ibacher commented on api/src/main/java/org/openmrs/module/emrapi/db/EmrApiDAOImpl.java:
```diff
@@ -55,4 +64,86 @@ public <T> List<T> executeHqlFromResource(String resource, Map<String, Object> p
+              .collect(Collectors.toList());
+      
+      // get encounter type ID for visit note, so that we don't have to run join query later
+      String hqlEncounterTypeId = "SELECT et.id FROM EncounterType et WHERE et.uuid = :encounterTypeUuid";
```
Change this to a join please.

jayasanka-sack commented:
cc: @denniskigen @gracepotma
